lib/mapObjects/CGHeroInstance.cpp

```diff
@@ -559,21 +559,10 @@ void CGHeroInstance::recreateSpecialtyBonuses(std::vector<HeroSpecial *> & speci
 
 void CGHeroInstance::updateSkill(SecondarySkill which, int val)
 {
+	removeBonuses(Selector::source(Bonus::SECONDARY_SKILL, which));
 	auto skillBonus = (*VLC->skillh)[which]->getBonus(val);
 	for (auto b : skillBonus)
-	{
-		// bonuses provided by different levels of a secondary skill are aggregated via max (not + as usual)
-		// different secondary skills providing the same bonus (e.g. ballistics might improve archery as well) are kept separate
```

Comment:
In past code added each level of secondary skills as separate bonus. So each level of skill might have completely different effects and all of them would be assigned to hero separately. Now on update you remove all assigned bonuses for the skill of hero, but only adding effects of single level.

So to make sure, there is no legacy data depended on old behavior? I suppose old saves loaded from legacy data are either converted by your compatibility code or simply didn't depend on it anyway?

Reply:
All secondary skill definitions specified the full bonus for each level anyways. The only potential issue is with mods that would assign different effects for each level (and assume them to be commulative) - but given that secondary skill modding is still fairly new, I doubt there are any.

Reply:
I guess fact that they're not cumulative give system even more flexibility so this is fine.